--- a/ssh-pkcs11.c	2015-01-21 14:57:43.000000000 -0800
+++ b/ssh-pkcs11.c	2015-01-21 15:05:23.000000000 -0800
@@ -509,10 +509,22 @@
 			if ((rsa = RSA_new()) == NULL) {
 				error("RSA_new failed");
 			} else {
+				CK_ULONG len = attribs[2].ulValueLen;
+				CK_CHAR_PTR ptr = attribs[2].pValue;
 				rsa->n = BN_bin2bn(attribs[1].pValue,
 				    attribs[1].ulValueLen, NULL);
-				rsa->e = BN_bin2bn(attribs[2].pValue,
-				    attribs[2].ulValueLen, NULL);
+				/*
+				 * Work-around for ActivIdentity bug: Public exponent
+				 * has leading zeros and one trailing zero.
+				 */
+ 				if(ptr[len - 1] == 0x0) {
+					len--;
+					while(ptr[0] == 0x0 && len > 1) {
+						ptr++;
+						len--;
+					}
+				}
+				rsa->e = BN_bin2bn(ptr, len, NULL);
 			}
 		} else {
 			cp = attribs[2].pValue;
