--- a/ssh-add.c	2017-12-29 00:24:45.000000000 -0800
+++ b/ssh-add.c	2017-12-29 00:28:46.000000000 -0800
@@ -337,12 +337,12 @@ add_file(int agent_fd, const char *filen
 }
 
 static int
-update_card(int agent_fd, int add, const char *id)
+update_card(int agent_fd, int add, const char *id, int no_pin)
 {
 	char *pin = NULL;
 	int r, ret = -1;
 
-	if (add) {
+	if (add && !no_pin) {
 		if ((pin = read_passphrase("Enter passphrase for PKCS#11: ",
 		    RP_ALLOW_STDIN)) == NULL)
 			return -1;
@@ -460,6 +460,7 @@ usage(void)
 	fprintf(stderr, "  -x          Lock agent.\n");
 	fprintf(stderr, "  -X          Unlock agent.\n");
 	fprintf(stderr, "  -s pkcs11   Add keys from PKCS#11 provider.\n");
+	fprintf(stderr, "  -n          Don't ask for PKCS#11 PIN.\n");
 	fprintf(stderr, "  -e pkcs11   Remove keys provided by PKCS#11 provider.\n");
 	fprintf(stderr, "  -q          Be quiet after a successful operation.\n");
 }
@@ -473,6 +474,7 @@ main(int argc, char **argv)
 	char *pkcs11provider = NULL;
 	int r, i, ch, deleting = 0, ret = 0, key_only = 0;
 	int xflag = 0, lflag = 0, Dflag = 0, qflag = 0;
+	int no_pin = 0;
 
 	ssh_malloc_init();	/* must be called before any mallocs */
 	/* Ensure that fds 0, 1 and 2 are open or directed to /dev/null */
@@ -500,7 +502,7 @@ main(int argc, char **argv)
 		exit(2);
 	}
 
-	while ((ch = getopt(argc, argv, "klLcdDxXE:e:qs:t:")) != -1) {
+	while ((ch = getopt(argc, argv, "klLcdDxXnE:e:qs:t:")) != -1) {
 		switch (ch) {
 		case 'E':
 			fingerprint_hash = ssh_digest_alg_by_name(optarg);
@@ -534,6 +536,9 @@ main(int argc, char **argv)
 		case 's':
 			pkcs11provider = optarg;
 			break;
+		case 'n':
+			no_pin = 1;
+			break;
 		case 'e':
 			deleting = 1;
 			pkcs11provider = optarg;
@@ -574,7 +579,7 @@ main(int argc, char **argv)
 	argc -= optind;
 	argv += optind;
 	if (pkcs11provider != NULL) {
-		if (update_card(agent_fd, !deleting, pkcs11provider) == -1)
+		if (update_card(agent_fd, !deleting, pkcs11provider, no_pin) == -1)
 			ret = 1;
 		goto done;
 	}
--- a/ssh-pkcs11.c	2017-12-29 00:24:45.000000000 -0800
+++ b/ssh-pkcs11.c	2017-12-29 00:28:46.000000000 -0800
@@ -232,7 +232,15 @@ pkcs11_find(struct pkcs11_provider *p, C
 int pkcs11_login(struct pkcs11_key *k11, CK_FUNCTION_LIST *f, struct pkcs11_slotinfo *si) {
 	char			*pin = NULL, prompt[1024];
 	CK_RV			rv;
-	if ((si->token.flags & CKF_LOGIN_REQUIRED) && !si->logged_in) {
+	if ((si->token.flags & CKF_LOGIN_REQUIRED) && !si->logged_in &&
+		(si->token.flags & CKF_PROTECTED_AUTHENTICATION_PATH)) {
+		if ((rv = f->C_Login(si->session, CKU_USER, NULL, 0)) == CKR_OK) {
+			si->logged_in = 1;
+		} else {
+			error("C_Login with protected authentication path failed: %lu", rv);
+		}
+	}
+    if ((si->token.flags & CKF_LOGIN_REQUIRED) && !si->logged_in) {
 		if (!pkcs11_interactive) {
 			error("need pin entry%s", (si->token.flags &
 			    CKF_PROTECTED_AUTHENTICATION_PATH) ?
@@ -527,10 +535,12 @@ pkcs11_open_session(struct pkcs11_provid
 	CK_FUNCTION_LIST	*f;
 	CK_SESSION_HANDLE	session;
 	int			login_required;
+	int			pap;
 
 	f = p->function_list;
 	login_required = p->slotinfo[slotidx].token.flags & CKF_LOGIN_REQUIRED;
-	if (pin && login_required && !strlen(pin)) {
+	pap = p->slotinfo[slotidx].token.flags & CKF_PROTECTED_AUTHENTICATION_PATH;
+	if (pin && login_required && !strlen(pin) && !pap) {
 		error("pin required");
 		return (-1);
 	}
@@ -540,7 +550,14 @@ pkcs11_open_session(struct pkcs11_provid
 		error("C_OpenSession failed: %lu", rv);
 		return (-1);
 	}
-	if (login_required && pin) {
+	if (login_required && pap) {
+		if ((rv = f->C_Login(session, CKU_USER, NULL, 0)) == CKR_OK) {
+			p->slotinfo[slotidx].logged_in = 1;
+		} else {
+			error("C_Login with protected authentication path failed: %lu", rv);
+		}
+	}
+	if (login_required && pin && !p->slotinfo[slotidx].logged_in) {
 		rv = f->C_Login(session, CKU_USER,
 		    (u_char *)pin, strlen(pin));
 		if (rv != CKR_OK && rv != CKR_USER_ALREADY_LOGGED_IN) {
